mojo/proposals/drain-collection-support.md

```diff
@@ -0,0 +1,452 @@
+
+    print('Testing set')
+    var set: Set[Foo] = {Foo(x="a"), Foo(x="b"), Foo(x="c")}
+    while set:
+        # Set.pop calls require raising context. Hashable items get copied
+        # since it copies the reference in the underlying `_DictKeyIter` 
+        try:
+            take_ownership(set.pop())
+        except:
+            print('Set is empty')
+
+    print('Testing dict')
+    var dict: Dict[String, Foo] = {"a": Foo(x="a"), "b": Foo(x="b"), 
+                                   "c": Foo(x="c")}
+    print('beginning iteration')
+    while dict:
+        try:
+            # Dict.popitem calls require raising context. 
+            # Executes a move, however must be in LIFO order.
+            var item = dict.popitem()
+            # DictEntry does not support move-based unpacking, so we need to 
+            # copy the key, then take the value.
+            take_ownership(item.key.copy(), item^.reap_value())
+        except:
+            print('Dict is empty')
+```
+
+This is made more complicated by combining with iter tools like `zip`:
+
+```mojo
+fn test_zip_from_2_lists():
+    var keys = List[String]("a", "b", "c")
+    var values = List[Foo](Foo(x="d"), Foo(x="e"), Foo(x="f"))
+    var it = zip(keys^, values^)
+    for item in it:
+        # Copies happen during zip iteration, then additionally
+        # must happen when passing to take_ownership. We are copying twice per 
+        # iter!
+        take_ownership(item[0].copy(), item[1].copy())
+```
+
+This is a common pattern, given a list of keys and a list of values, we want to
+zip them together into a map-like object (Dict, Tree, etc.).
+
+### How do other languages handle this?
+
+Rust
+
+> TODO: Would like input from people with Rust heavy backgrounds to clarify this
+> is how Rust drain actually works, particularly "when do the source collections
+> get cleared"?
+
+Rust has a struct called
+[`vec/Drain`][2] and
+[`hashmap/Drain`][3]
+that is used by collections like `Vec` and `HashMap` to drain the collection.
+
+[Vec.drain](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.drain)
+
+```rust
+let mut v = vec![1, 2, 3];
+let u: Vec<_> = v.drain(1..).collect();
+assert_eq!(v, &[1]);
+assert_eq!(u, &[2, 3]);
+
+// A full range clears the vector, like `clear()` does
+v.drain(..);
+assert_eq!(v, &[]);
+```
+
+[HashMap.drain](https://doc.rust-lang.org/std/collections/struct.HashMap.html#method.drain)
+
+```rust
+use std::collections::HashMap;
+
+let mut a = HashMap::new();
+a.insert(1, "a");
+a.insert(2, "b");
+
+for (k, v) in a.drain().take(1) {
+    assert!(k == 1 || k == 2);
+    assert!(v == "a" || v == "b");
+}
+
+assert!(a.is_empty());
+```
+
+For sequential collections like Vec:
+
+- `drain(..)` clears the full collection either on full iteration or when the
+iterator is dropped (destroyed).
+- `drain(start..end)` drains a range of elements either on full iteration or when
+the iterator is dropped (destroyed).
+
+For hashable collections like HashMap:
```

Comment:
Yep! Or for non-sequential/unordered collections where taking a specific "range" doesn't really make sense - their drain function won't support a range (synonymous to Mojo's slice)